docs/tracing.md

```diff
@@ -0,0 +1,74 @@
+# Tracing
+
+## Introduction
+
+Instrumentation based tracing is implemented with
+[`log`](https://github.com/rust-lang/log) and
+[`log_instrument`](https://github.com/JonathanWoollett-Light/log-instrument),
+outputting a `Trace` level log when entering and exiting every function.
+
+It is disabled by default at compile-time, as such the
+**tracing functionality has no impact on the release binary.**
+
+To enable tracing, remove `"max_level_debug"` from the features
+for `log` under [`../src/vmm/Cargo.toml`](../src/vmm/Cargo.toml):
+
+```diff
+- log = { version = "0.4.17", features = ["std", "serde", "max_level_debug"] }
++ log = { version = "0.4.17", features = ["std", "serde"] }
+```
+
+This will result in a massive increase in the binary size and a
+massive regression in performance. To mitigate the performance affects
+you can use filtering.
+
+## Filtering
+
+### Run-time
+
+You can filter by module and/or file at runtime, e.g.:
+
+```bash
+curl -X PUT --unix-socket "${API_SOCKET}" \
+    --data "{
+        \"level\": \"Trace\",
+        \"file\": \"src/firecracker/src/api_server_adapter.rs\",
+    }" \
+    "http://localhost/logger"
+```
+
+- Instrumentation logs are `Trace` level, at runtime the level must be
+  set to `Trace` to see them.
+- This will only output logs from `src/firecracker/src/api_server_adapter.rs`.
+
+This will mitigate most of the performance regression.
+
+### Compile-time
+
+It can be valuable to produce a binary with targetted tracing that does
+not need to be filtered at run-time.
```

Comment:
I'd start with a stronger motivation here:

"If the performance impact of runtime filtering is too significant, it is possible to statically exclude specific traces. [...]"

Reply:
I've re-worded it:
```
Specific environments can restrict run-time configuration, in these
environments it becomes necessary to support targeted tracing without
run-time re-configuration, for this compile-time filtering must be used.
```